bpf/lib/nat.h

```diff
@@ -1649,6 +1731,72 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
 			tuple.dport = icmp6hdr.icmp6_dataun.u_echo.identifier;
 			tuple.sport = 0;
 			break;
+		case ICMPV6_PKT_TOOBIG: {
+			struct ipv6hdr iphdr;
+			__be16 identifier;
+			__u8 type;
+			__u32 icmpoff = off;
+
+			/* According to the RFC 5508, any networking
+			 * equipment that is responding with an ICMP Error
+			 * packet should embed the original packet in its
+			 * response.
+			 */
+
+			/* Note related to how is computed the offset. The
+			 * ICMPV6_PKT_TOOBIG does not include identifer and
+			 * sequence in its headers.
+			 */
+			icmpoff += sizeof(icmp6hdr) - sizeof(icmp6hdr.icmp6_dataun.u_echo);
+
+			if (ctx_load_bytes(ctx, icmpoff, &iphdr,
+					   sizeof(iphdr)) < 0)
+				return DROP_INVALID;
+
+			/* From the embedded IP headers we should be able
+			 * to determine corresponding protocol, IP src/dst
+			 * of the packet sent to resolve the NAT session.
+			 */
+
+			tuple.nexthdr = iphdr.nexthdr;
+			ipv6_addr_copy(&tuple.saddr, (union v6addr *)&iphdr.daddr);
+			ipv6_addr_copy(&tuple.daddr, (union v6addr *)&iphdr.saddr);
+
+			icmpoff += ipv6_hdrlen(ctx, &iphdr.nexthdr);
```

Comment:
I don't think this will work? `ipv6_hdrlen()` loads from `ctx + ETH_HLEN`, but you want to access the inner headers here. I'd suggest to introduce a `l3_off` parameter.

Reply:
I should have fixed it. If you can double check.

Reply:
Please have a closer look.

`ipv6_hdrlen()` needs to walk the packet's Extension headers. But if the inner packet contains **any** relevant extension header (based on `ip6->nexthdr`), then `ipv6_hdrlen()` will currently try to load that extension header from `ctx->data + ETH_LEN + sizeof(struct ipv6hdr)`. So you end up accessing the outer ICMP header instead.

Reply:
I opened a PR that gives ability to set an offset to `ipv6_hdrlen`, function called `ipv6_hdrlen_from` I hope that makes sense.

Reply:
:+1: Please just include that patch in this PR.

Reply:
OK I have cherry-picked the commit in this PR and used the new function. Usage of this new function simplified computing the lentgh :-)